File: ScanProject/PyAnalysis/SonarPy/Medium/633-multiply-strings.py
#https://leetcode.com/problems/multiply-strings/
#Given two non-negative integers num1 and num2 represented as strings, return the product of num1 and num2, also represented as a string.
#Note: You must not use any built-in BigInteger library or convert the inputs to integer directly.


class Solution(object):
    def multiply(self, num1, num2):
        """
        :type num1: str
        :type num2: str
        :rtype: str
        """ 
              
        if num1=="0" or num2=="0":
            return "0"
        if num1=="1":
            return num2
        if num2=="1":
            return num1
        if len(num1)<len(num2):
            num1,num2=num2,num1
        num1=num1[::-1]
        num2=num2[::-1]
        res="0"
        for i in range(len(num2)):
            res=self.add(res,self.multiplyOne(num1,num2[i],i))
        return res
    def add(self,num1,num2):
            res=""
            i,j,carry=len(num1)-1,len(num2)-1,0
            while i>=0 or j>=0 or carry:
                if i>=0:
                    carry+=int(num1[i])
                    i-=1
                if j>=0:
                    carry+=int(num2[j])
                    j-=1
                res=str(carry%10)+res
                carry=carry//10
            return res
    def multiplyOne(self,num1,num2,i):
            res=""
            carry=0
            for j in range(len(num1)):
                carry+=int(num1[j])*int(num2)
                res=str(carry%10)+res
                carry=carry//10
            if carry>0:
                res=str(carry)+res
            res=res+"0"*i
            return res

        

    

File: ScanProject/JSAnalysis/SonarJS/Medium/633-multiply-strings.js
/**https://leetcode.com/problems/multiply-strings/ */
//Given two non-negative integers num1 and num2 represented as strings, return the product of num1 and num2, also represented as a string.
//Note: You must not use any built-in BigInteger library or convert the inputs to integer directly.


/**
 * @param {string} num1
 * @param {string} num2
 * @return {string}
 */
var multiply = function(num1, num2) {
    let result = [];
    let carry = 0;
    for(let i = num1.length - 1; i >= 0; i--){
        let current = [];
        for(let j = num2.length - 1; j >= 0; j--){
            let product = (num1[i] - '0') * (num2[j] - '0') + carry;
            carry = Math.floor(product / 10);
            current.unshift(product % 10);
        }
        if(carry > 0){
            current.unshift(carry);
            carry = 0;
        }
        for(let k = 0; k < num1.length - 1 - i; k++){
            current.push(0);
        }
        result = add(result, current);
    }
    return result.join('');
    
    
};
function add(num1, num2){
    let result = [];
    let carry = 0;
    let i = num1.length - 1;
    let j = num2.length - 1;
    while(i >= 0 && j >= 0){
        let sum = num1[i] + num2[j] + carry;
        carry = Math.floor(sum / 10);
        result.unshift(sum % 10);
        i--;
        j--;
    }
    while(i >= 0){
        let sum = num1[i] + carry;
        carry = Math.floor(sum / 10);
        result.unshift(sum % 10);
        i--;
    }
    while(j >= 0){
        let sum = num2[j] + carry;
        carry = Math.floor(sum / 10);
        result.unshift(sum % 10);
        j--;
    }
    if(carry > 0){
        result.unshift(carry);
    }
    return result;
}
    